draft-ietf-lamps-dilithium-certificates.md

```diff
@@ -431,10 +432,72 @@ OneAsymmetricKey is replicated below.
   2021 ASN.1 syntax {{X680}}.
 </aside>
 
-When used in a OneAsymmetricKey type, the privateKey OCTET STRING contains
-the raw octet string encoding of the 32-octet seed. The publicKey field
-SHOULD be omitted because the public key can be computed as noted earlier
-in this section.
+For ML-DSA private keys, the privateKey field in OneAsymmetricKey contains one of
+the following CHOICE structures encoded as an OCTET STRING. The seed format is a
+fixed 32 bytes for all security levels, while the expandedKey and both formats
+vary in size by security level:
+
+
+~~~
+ML-DSA-44-PrivateKey ::= CHOICE {
+  seed [0] OCTET STRING (SIZE (32)),
+  expandedKey OCTET STRING (SIZE (2560)),
+  both SEQUENCE {
+      seed OCTET STRING (SIZE (32)),
+      expandedKey OCTET STRING (SIZE (2560))
+      }
+  }
+~~~
+
+~~~
+ML-DSA-65-PrivateKey ::= CHOICE {
+  seed [0] OCTET STRING (SIZE (32)),
+  expandedKey OCTET STRING (SIZE (4032)),
+  both SEQUENCE {
+      seed OCTET STRING (SIZE (32)),
+      expandedKey OCTET STRING (SIZE (4032))
+      }
+  }
+~~~
+
+~~~
+ML-DSA-87-PrivateKey ::= CHOICE {
+  seed [0] OCTET STRING (SIZE (32)),
+  expandedKey OCTET STRING (SIZE (4896)),
+  both SEQUENCE {
+      seed OCTET STRING (SIZE (32)),
+      expandedKey OCTET STRING (SIZE (4896))
+      }
+  }
+~~~
+
+The CHOICE allows three representations of the private key:
+
+1. The seed format (tag `[0]`) contains just the 32-byte seed value (xi)
+   from which both the expanded private key and public key can be derived
+   using ML-DSA.KeyGen_internal(xi).
+
+2. The expandedKey format contains the full expanded private key that was
+   derived from the seed.
+
+3. The both format contains both the seed and expanded key, allowing for
+   validation that the expanded key was correctly derived from the seed.
+
+When encoding an ML-DSA private key in a OneAsymmetricKey object, any of
+these three formats may be used, though the seed format is RECOMMENDED
+for storage efficiency.
+
+The "privateKeyAlgorithm" field uses the AlgorithmIdentifier structure with
+the appropriate OID as defined in Section 3. If present, the "publicKey"
+field will hold the encoded public key as defined in Section 5.
+
+NOTE: While the private key can be stored in multiple formats, the seed-only
+format is RECOMMENDED as it is the most compact representation. Both the
+expanded private key and the public key can be deterministically derived
+from the seed using ML-DSA.KeyGen_internal(xi). While the publicKey field
+and expandedKey format are technically redundant when using the seed-only
+format, they MAY be included to enable keypair consistency checks during
+import operations.
```

Comment:
Same issue here.  The first sentence is correct (more compact representation), but what follows again suggests that the public key can only be recovered from the seed, which is NOT the case. OpenSSL, for example, always recovers the public key from the private alone when importing ML-DSA private keys (if a public key is also supplied on import it is compared with the generated value and the pair reject if it does not match).

Reply:
Agree here also, similarly, AWS-LC does the same, we generate the public key based on the provided private key. Again, the sentence you interpret differently to me is simply pointed out that the seed _can_ provide both pub and priv, not that it can only be recovered.

I was actually going to propose a section in the draft that speaks to how to generate pk from sk, given that this isn't provided in the reference implementation. I mentioned that in this issue: https://github.com/lamps-wg/dilithium-certificates/issues/77. Perhaps such an inclusion can then be referred to in the text you point out here and above, to indicate that the pk can be constructed from the sk. I think that should help satisfy your remarks.

Reply:
Well, just a subset (4–8) of the Algorithm 6 ML-DSA.KeyGen_internal(𝜉) steps, since the private vectors are generated, and then the public key just from those.